[DROOLS-7120] Drools 8 docs : Rule Unit API (#4691)
kiegroup/drools: drools-docs/src/modules/ROOT/pages/KIE/BuildDeployUtilizeAndRun/_rule-unit-api.adoc

[id='rule-unit-api_{context}']

= Rule Unit API

Previous sections in this chapter explained traditional KIE APIs to use {RULE_ENGINE}. However, as introduced in xref:getting-started/index.adoc#first-rule-project_getting-started[First Rule Project], Rule Unit is a new and recommended style for implementing rules in {PRODUCT} 8.

> A rule unit is an atomic module defining a set of rules and a set of strongly typed *data sources* through which the facts processed by the rules are inserted. *data sources* are of 2 kinds: `DataStream` and `DataStore` that will be described later in this section.

Reviewer: ```suggestion
A rule unit is an atomic module defining a set of rules and a set of strongly typed *data sources* through which the facts processed by the rules are inserted. The *data sources* of 2 kinds: `DataStream` and `DataStore` which will be described later in this section.
```

image::KIE/BuildDeployUtilizeAndRun/RuleUnit01.png[align="center"]

> `DataSources` could be shared, so coordinate rule units.

Reviewer 2: ```suggestion
`DataSources` can be shared across different units, providing a coordination mechanism between them.
```

image::KIE/BuildDeployUtilizeAndRun/RuleUnit02.png[align="center"]

== RuleUnitData

`RuleUnitData` is an interface to define your Rule Unit. Rule unit implementation would be like this:

[source,java]
----
public class MeasurementUnit implements RuleUnitData {

    private final DataStore<Measurement> measurements;
    private final Set<String> controlSet = new HashSet<>();

    public MeasurementUnit() {
        this(DataSource.createStore());
    }

    public MeasurementUnit(DataStore<Measurement> measurements) {
        this.measurements = measurements;
    }

    public DataStore<Measurement> getMeasurements() {
        return measurements;
    }

    public Set<String> getControlSet() {
        return controlSet;
    }
}
----

In this case, `Measurement` is your fact class, so you need to implement it as well.

> The name of the rule unit class `MeasurementUnit` is linked to DRL rules using `unit`.

Reviewer 2: ```suggestion
The name of the rule unit class `MeasurementUnit` is linked to DRL rules using the `unit` statement.
```

----

* `DataStore`: A writable storage option for adding or removing data and then notifying all subscribers that mutable data has been modified. Rules can pattern-match against incoming values and update or remove available values.
For users familiar with traditional DRL syntax, this option is equivalent to a typed version of an entry point. In fact, a `DataStore<Object>` is equivalent to a traditional-style entry point.
+
.Example DataStore data source usage
[source,java]
----
DataStore<Measurement> measurements = DataSource.createStore();
Measurement measurement = new Measurement("color", "red");
// Add value and notify all subscribers
DataHandle mHandle = measurements.add(measurement);
measure.setValue("blue");
// Notify all subscribers that the value referenced by `mHandle` has changed
measurements.update(mHandle, measurement);
// Remove value referenced by `mHandle` and notify all subscribers
measurements.remove(mHandle);
----

* `SingletonStore`: A writable storage option for setting or clearing a single element and then notifying all subscribers that the element has been modified. Rules can pattern-match against the value and update or clear available values.

> For users familiar with traditional DRL syntax, this option is equivalent to a global. In fact, a `Singleton<Object>` is similar to a traditional-style global, except that when used in conjunction with rules, you can pattern-match against it.

Reviewer 2: ```suggestion
For users familiar with traditional DRL syntax, this option is similar to a global but reactive to changes. In fact, a `Singleton<Object>` is similar to a traditional-style global, except that when used in conjunction with rules, you can pattern-match against it.
```

if it is _confirmed_ it is reactive to changes, I believe _similar_ is more appropriate here my2c (iirc global is not reactive to changes)

Author: Thanks. You are right.

> For users familiar with traditional DRL syntax, this option is similar to a global but reactive to changes. In fact, a `Singleton<Object>` is similar to a traditional-style global, except that when used in conjunction with rules, you can pattern-match against it.

Reviewer: ```suggestion
For users familiar with traditional DRL syntax, this option is similar to a global but reactive to changes. A `Singleton<Object>` is similar to a traditional-style global, except that when used in conjunction with rules, you can pattern-match against it.
```

+
.Example SingletonStore data source definition
[source,java]
----
SingletonStore<Measurement> measurement = DataSource.createSingleton();
Measurement m1 = new Measurement("color", "red");
// Add value `m1` and notify all subscribers
measurement.set(m1);
measure.setValue("blue");
// Notify all subscribers that the value has changed
measurement.update();

Measurement m2 = new Measurement("color", "green");
// Overwrite contained value with `m2` and notify all subscribers
measurement.set(m2);
measure2.setValue("black");
// Notify all subscribers that the value has changed
measurement.update();

// Clear store and notify all subscribers
measurement.clear();
----

Subscribers to a data source are known as _data processors_. A data processor implements the `DataProcessor<T>` interface. This interface contains callbacks to all the events that a subscribed data source can trigger.

.DataProcessor
[source,java]
----
public interface DataProcessor<T> {
    default void insert(T object) {
        insert(null, object);
    }
    FactHandle insert(DataHandle handle, T object);
    void update(DataHandle handle, T object);
    void delete(DataHandle handle);
}
----

> The `DataHandle` is an internal reference to an object of a data source. Each callaback method might or might not be invoked, depending on whether the corresponding data source implements the capability. For example, a `DataStream` source invokes only the `insert` callback, whereas a `SingletonStore` source invokes the `insert` callback on `set` and the `delete` callback on `clear` or before an overwriting `set`.

Reviewer: ```suggestion
The `DataHandle` is an internal reference to an object of a data source. Each callback method might or might not be invoked, depending on whether the corresponding data source implements the capability. For example, a `DataStream` source invokes only the `insert` callback, whereas a `SingletonStore` source invokes the `insert` callback on `set` and the `delete` callback on `clear` or before an overwriting `set`.
```
